### ISSUE TITLE
Downloading data for multiple clans

### ISSUE THREAD

**Issue by mikewojtkiewicz** [Bungie-net/api#606]
I have a unique situation that I'm working with, and I'm not sure of the best solution. I'm working on downloading data for 7 different clans that are all apart of one Discord server, and we need to use the api to view individual members activity history.  The issue however is that in order to do this, I need to make thousands of calls to the api and store it in a database to be queried and used later. Everything I've done to date works fine, but the full activity check loop takes almost a full day to complete. Is there a better option that I'm not thinking of? I'm also worries about being rate limited by Bungie since I'm making so many calls. 

Edit: I should also mention the reason this takes so long is because after I get an activity id, I also need to go into the pgcr and get the list of membership ids from there and do another call to see if the member played with another member of one of the 7 clans.

**Comment by vthornheart-bng** [comment]
For the PGCR portion of the querying, is it that you need to know if they played with a fellow clan member that week (i.e. is the app tracking this purely for in-game Clan Rewards purposes), or are you doing a breakdown of the specific match etc?  For instance, if we found a way to return you a bit of information about whether or not you have played with a fellow clan member this week and returned that, would it eliminate your need to call the stats API?

I don't actually know offhand if we have that kind of information available, but if that would be a helpful alternative I can take a look and see if we know that information anywhere other than in the PGCRs.

**Comment by mikewojtkiewicz** [comment]
That's a little different than what I'm thinking about, actually. It could be useful info, but it wouldn't save me time because I need to check if each member id exists in a set list of group ids (there are 7 individual clans). The only thing that I think would speed my process up is bringing back the clan alliance feature. But I understand the reason for why you haven't brought that back.

**Comment by migit128** [comment]
I do **very** similar things, but it doesn't take me anywhere near as long to get my data.  If I clear out my db entirely it'd take about an hour to download every single person's full activity history for all active/deleted characters, compute KDs for overall/Comp playlists, get easier to query raid completions, and more.

To update the character list in the clan, it takes 3.5 minutes.  To update everyone's activity history, it takes 4 minutes.  This is with 550ish clan members.

I deleted all my API calls getting aggregate stats and PVP KDs because I realized I had all that data in the raw form.  If you store the kills/assists/deaths/game mode in your db's activity history table, it's trivial to compute them.  Same goes for raid completions.  Keep a list of hashes for each raid type in your db and you can use that to count up their raid clears for each raid.

I don't bother getting PGCRs atm because I store the instance id in my activity table.  All you need to do to see if they played with clan members is group things up by that ID.  If it's in there more than once they played with someone in the clan. 

Thorn told me the rate limit is 25/second.  If you're in a section of code where the calls go fast, don't thread it.  Activity history calls take a while though, so I have 10 threads running when I'm grabbing that data.  My account/character data calls go faster, so only 5 threads.  If you're running on a VM in the cloud, make sure you have a private public IP (got far less errors after I got my own public IP).  If you start threading your calls though you need to make sure your error handling is on point, because you WILL get more errors.

Simply put, API calls are expensive.  Finding ways to minimize your API calls is the main way you'll speed up your run times.

Also, what's your clan?

**Comment by mikewojtkiewicz** [comment]
Well, I mostly solved the lagging issue I had by first, instead of inserting a row into the db on every api call, I now store all the data in memory, then perform a bulk insert. Second, I still get the pgcr because I need to have time spent in activity as well as the completed status. Doing these things saved me quite a bit of time. It still take a while, but I also have around 3 to 4 threads running now myself. So it'll complete the full thing in around 5 hours. Still quite some time, but a lot better than it was before! 

Thanks for the comment though!

**Comment by migit128** [comment]
You dont need the PGCR for time spent in an activity.  The activity history has start/end times for the user and the complete activity.  There is a completed flag in the activity as well.  Does the PGCR give more info on how it was completed?

![This is the data I'm getting from the activity endpoint (and a little more)](https://user-images.githubusercontent.com/7931231/44605335-c022a480-a79d-11e8-91cf-035203dd2c7c.png)

**Comment by mikewojtkiewicz** [comment]
Did I miss that? I don't see it in the endpoint. This is the one I'm using: https://bungie-net.github.io/multi/operation_get_Destiny2-GetActivityHistory.html#operation_get_Destiny2-GetActivityHistory

**Comment by migit128** [comment]
Thats the right endpoint.  (sample URL: https://www.bungie.net/platform/Destiny2/4/Account/4611686018475037779/Character/2305843009357183996/Stats/Activities/?count=250&mode=0&page=0)


The "activities" section it has the period variable.  that's the start time of the event.

In the values section:
timePlayedSeconds - duration the user was in there.
startSeconds - seconds that passed since the start of the event before this user entered.
activityDurationSeconds - complete activity duration

With those 3 values and the start time you can figure out the time the user entered, the time the user left, the start time of the event, and the end time of the event.

Completed flag - 0 or 1 if they complete it.  somethings like patrols never show as completed.  The completion reason I never really looked into much.  I dont use it in my code.

You can then get kills/deaths/assists/etc from the other variables of the values section

```
             "assists": {
                        "statId": "assists",
                        "basic": {
                            "value": 114,
                            "displayValue": "114"
                        }
                    },
                    "completed": {
                        "statId": "completed",
                        "basic": {
                            "value": 0,
                            "displayValue": "No"
                        }
                    },
                    "deaths": {
                        "statId": "deaths",
                        "basic": {
                            "value": 16,
                            "displayValue": "16"
                        }
                    },
                    "kills": {
                        "statId": "kills",
                        "basic": {
                            "value": 202,
                            "displayValue": "202"
                        }
                    },
                    "opponentsDefeated": {
                        "statId": "opponentsDefeated",
                        "basic": {
                            "value": 316,
                            "displayValue": "316"
                        }
                    },
                    "efficiency": {
                        "statId": "efficiency",
                        "basic": {
                            "value": 19.75,
                            "displayValue": "19.75"
                        }
                    },
                    "killsDeathsRatio": {
                        "statId": "killsDeathsRatio",
                        "basic": {
                            "value": 12.625,
                            "displayValue": "12.63"
                        }
                    },
                    "killsDeathsAssists": {
                        "statId": "killsDeathsAssists",
                        "basic": {
                            "value": 16.1875,
                            "displayValue": "16.19"
                        }
                    },

   "activityDurationSeconds": {
                        "statId": "activityDurationSeconds",
                        "basic": {
                            "value": 5712,
                            "displayValue": "1h 35m"
                        }
                    },
                    "completionReason": {
                        "statId": "completionReason",
                        "basic": {
                            "value": 0,
                            "displayValue": "Objective Completed"
                        }
                    },
  "startSeconds": {
                        "statId": "startSeconds",
                        "basic": {
                            "value": 0,
                            "displayValue": "0m 0s"
                        }
                    },
                    "timePlayedSeconds": {
                        "statId": "timePlayedSeconds",
                        "basic": {
                            "value": 4886,
                            "displayValue": "1h 21m"
                        }
                    },
```
